core/src/style/style.cpp

```diff
@@ -68,57 +75,110 @@ void Style::setLightingType(LightingType _type){
 
 }
 
-void Style::addLayer(const std::pair<std::string, StyleParamMap>&& _layer) {
+void Style::addLayer(std::shared_ptr<SceneLayer> _layer) {
 
     m_layers.push_back(std::move(_layer));
 
 }
 
+void Style::applyLayerFiltering(const Feature& _feature, const Context& _ctx, std::bitset<MAX_LAYERS>& _uniqueID,
+                                   StyleParamMap& _styleParamMapMix, std::weak_ptr<SceneLayer> _uberLayer) const {
+
+    std::vector<std::weak_ptr<SceneLayer>> sLayers;
+    sLayers.reserve(_uberLayer.lock()->getSublayers().size() + 1);
+    sLayers.push_back(_uberLayer);
+
+    auto sLayerItr = sLayers.begin();
+
+    //A BFS traversal of the SceneLayer graph
+    while (sLayerItr != sLayers.end()) {
+
+        auto sceneLyr = (*sLayerItr).lock();
+
+        if ( sceneLyr->getFilter()->eval(_feature, _ctx)) { // filter matches
+
+            _uniqueID.set(sceneLyr->getID());
+
+            if(s_styleParamMapCache.find(_uniqueID) != s_styleParamMapCache.end()) {
+
+                _styleParamMapMix = s_styleParamMapCache.at(_uniqueID);
+
+            } else {
+
+                /* update StyleParam with subLayer parameters */
+                auto& sLayerStyleParamMap = sceneLyr->getStyleParamMap();
+                for(auto& styleParam : sLayerStyleParamMap) {
+                    auto it = _styleParamMapMix.find(styleParam.first);
+                    // C++17 has an insert_or_assign which does the same thing.
+                    if(it != _styleParamMapMix.end()) {
+                        it->second = styleParam.second;
+                    } else {
+                        _styleParamMapMix.insert(styleParam);
+                    }
```

Comment:
How is this different from

    _styleParamMapMix[styleParam.first] = styleParam.second; 

The separate insert would only save a copy if styleParam could be moved.